README.md

```diff
@@ -297,7 +297,27 @@ int7 = 1_2_3_4_5     # VALID but discouraged
 ```
 
 Leading zeros are not allowed. Integer values `-0` and `+0` are valid and
-identical to an unprefixed zero. Hex, octal, and binary forms are not allowed.
+identical to an unprefixed zero.
+
+Integer values may also be expressed in hexadecimal, octal, or binary. Hex
+values are case insensitive. Leading zeros are allowed. Underscores are allowed
+between digits (but not between the prefix and the value).
+
+```toml
+# hexadecimal with prefix `0x`
+hex1 = 0xDEADBEEF
+hex2 = 0xdeadbeef
+hex3 = 0xdead_beef
+hex4 = +0xDEADBEEF # same as above
+hex5 = -0xDEADBEEF # negative of above
```

Comment:
What are the semantics of negative hex literals if the literal is large enough to have its 63 bit set? E.g., since integer literals are considered (a couple lines further down in the spec) to be signed longs (signed 64-bit integers), what are the semantics of `0x8000_0000_0000_0000` or `0xffff_ffff_ffff_ffff`? And what about the semantics of `-0x8000_0000_0000_0000` and `-0xffff_ffff_ffff_ffff`?

Most people will probably implement this spec such that `0xffff_ffff_ffff_ffff` becomes -1, and thus `-0xffff_ffff_ffff_ffff` will end up producing the value 1. And then both `0x8000_0000_0000_0000` and `-0x8000_0000_0000_0000` will end up producing the value -9,223,372,036,854,775,808 (which is the corner case of 64-bit int negation via two's complement), whereas `0x8000_0000_0000_0001` will produce -9,223,372,036,854,775,807 and thus `-0x8000_0000_0000_0001` will produce +9,223,372,036,854,775,807.

I think it might be wise to document those specific corner cases, either a few examples in the spec or with a few test cases to ensure that all implementations handle them identically.

Reply:
@rmunn How about letting negative number follows similar "loosen-later" policy? What't the use case of negative hexadecimal or binary?

Reply:
That's one of the things I've been wondering, and my original proposal was to disallow negative hex numbers since their semantics aren't totally clear in edge cases. (And languages vary about whether they're allowed: for example, in F#, `0x8000000000000000L` (the L is required in F# and means "this is a `long` literal) is -9,223,372,036,854,775,808 (and is of type "signed long"), and `-0x8000000000000000L` is *also* -9,223,372,036,854,775,808 since that's the one bit pattern that is the two's complement of itself. Java appears to follow the same semantics as F#, at least if http://www.javarepl.com/ is to be trusted. However, in C#, `0x8000000000000000` is *positive* 9,223,372,036,854,775,808 (and is of type "unsigned long"), and `-0x8000000000000000` produces `error CS0023: The '-' operator cannot be applied to operand of type 'ulong'`.

So since there's not one consistent way to handle negative hex numbers among major languages, my original proposal was to forbid them in TOML. However, @mojombo is the one whose name is going on the language, so it's ultimately his call. Since he wants to allow negative hex literals, I won't argue with that decision, and I'll just point out corner cases that need to be covered, at least in the tests if not in the spec.

Reply:
All fair points. I originally allowed it for consistency sake, but you're right, I don't see any compelling use cases, and it opens up the can of worms you've stated. I just pushed a commit that revises this proposal to disallow signed hex/oct/bin values.